### ISSUE TITLE
Slow Evaluation for imported Keras model

### ISSUE THREAD

**Issue by bozhou1990** [deeplearning4j/deeplearning4j#4010]
#### Issue Description
I imported a Keras LSTM model using DL4J. Our use case is to train LSTM models using Keras on Tensorflow and use DL4J to import the model for online real-time evaluation. DL4J's evaluation latency is significantly slower than Keras.

Here is the structure of our LSTM model:
    
    model1 = Sequential()
    model1.add(LSTM(256, input_shape=(120, 33), return_sequences=True))
    model1.add(Dropout(0.5))
    model1.add(LSTM(256))
    model1.add(Dropout(0.5))
    model1.add(Dense(256, activation='tanh'))
    model1.add(Dropout(0.5))
    model1.add(Dense(128, activation='relu'))
    model1.add(Dropout(0.5))
    model1.add(Dense(num_classes))
    model1.add(Activation('softmax'))
    return model1

Simply put, it is a sequence model with 2 LSTM layers followed by 3 dense layers. It uses 33 features. Sequence length is 120. At evaluation time, each sequence is generated individually. In other words, the mini-batch size is always 1. Our use case requires an evaluation latency of under 100ms at p99 level.

Please describe our issue, along with:
- expected behavior
Keras evaluation latency is generally between 30ms and 40ms.
- encountered behavior
DL4J evaluation latency is generally above 100ms. I gathered the following statistics over 100 samples.
Average: 124ms
p50: 119ms
p90: 145ms
p99: 189ms
Max: 268ms

#### Version Information
Please indicate relevant versions, including, if relevant:

* Deeplearning4j version
0.9.1
* platform information (OS, etc)
Ubuntu 16.04 LTS, Linux 4.10, running on a virtual box with 2 vCPUs
* CPU backend
nd4j-native
* Keras
Keras 1.2.2, tensorflow 1.2.1, numpy 1.13.1

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by AlexDBlack** [comment]
> use DL4J to import the model for online real-time evaluation

For "evaluation" do you simply mean generating predictions - or do you mean "evaluation" as in accuracy/f1 calculation, etc?
Can you show us the code for this?

**Comment by bozhou1990** [comment]
Evaluation here means prediction only.

My DL4J code looks something like this:

for (String record : records) {
    final INDArray input = parseNumpyRecord(record); // Return INDArray with shape (1, FEATURE_NUM, SEQUENCE_LEN)
    final long start = System.currentTimeMillis();
    final INDArray output = network.output(input);
    final long end = System.currentTimeMillis();
    latencies.add(end - start);
}

**Comment by raver119** [comment]
Show your import code please. Also, what happens when you enable single workspace for inference?

**Comment by bozhou1990** [comment]
Import code:

MultiLayerNetwork network = KerasModelImport
            .importKerasSequentialModelAndWeights(
                "LSTM_keras122.h5",
                false);

I was not aware of DL4J's workspace feature. The data in the summary was obtained without using workspace. I can try it to see whether I can better latency results.

**Comment by bozhou1990** [comment]
I tried using single workspace for inference. The latency barely changed. Here is my test code:

        MultiLayerNetwork kerasModel = KerasModelImport
            .importKerasSequentialModelAndWeights(
                "/tmp/LSTM_keras122.h5",
                false);

        MultiLayerConfiguration config = kerasModel.getLayerWiseConfigurations().clone();
        config.setInferenceWorkspaceMode(WorkspaceMode.SINGLE);
        MultiLayerNetwork model = new MultiLayerNetwork(config, kerasModel.params());

        List<Long> latencies = new ArrayList<>();

        try (final BufferedReader reader = Files.newBufferedReader(Paths.get("/tmp/record1.txt"))) {
            String record = null;
                while ((record = reader.readLine()) != null) {
                    final INDArray input = parseNumpyRecord(record);
                    final long start = System.currentTimeMillis();

                    INDArray output = model.output(input);

                    final long end = System.currentTimeMillis();
                    latencies.add(end - start);
                    printResult(output);
                }
        }

**Comment by bozhou1990** [comment]
Hi guys,

Any update on this issue?

**Comment by kyoungah-kim** [comment]
Hi, any good news on this issue? I'm experiencing similar issue with Linux machine.

**Comment by maxpumperla** [comment]
@bozhou1990 @kyoungah-kim I'm still puzzled how this would have anything to do with keras import. Are you saying that if you build the same model in purely dl4j it's faster? (I doubt that) Or are you just generally not happy with inference speed?

@raver119 do you have an idea how to help these guys?

**Comment by kyoungah-kim** [comment]
What is expected inference speed comparing to Keras? My model (*.h5) prediction speed is around <15ms in average on Keras while it becomes 120ms when I load the model into Dl4j (version 0.9.2 snapshot). I am just curious about expected performance :$  
Thanks!!

**Comment by raver119** [comment]
What’s your model can i see gist with config?
Also, what was the compilation mode for libnd4j? avx/avx2 enabled? 
And workspaces - were they enabled?

**Comment by kyoungah-kim** [comment]
it's LSTM model, workspaces have been enabled as single mode. I didn't use compilation extension and just rebuilt libnd4j with avx2 and looks like speed is about 1.5x faster now, thanks! Is there any other possible optimization? I'm using Haswell processor so avx2 might be the best available extension.

**Comment by maxpumperla** [comment]
I'm closing this for now, as I don't see how we could end up with a satisfying result. Imported models are plain DL4J models and they are as fast or slow as they currently are. There are quite a few knobs to tweak, but I'm not sure how to address all concerns in this issue.

Feel free to re-open if I miss something.

**Comment by whaozl** [comment]
> parseNumpyRecord

@bozhou1990 can you show this parseNumpyRecord method code?

thanks!

**Comment by bozhou1990** [comment]
Sorry. I no longer have access to that code.

**Comment by agibsonccc** [comment]
@whaozl please open new issues if you have them.  If you want to use numpy, we have several methods for reading numpy binary. Open a new issue with what you are trying to do or ask in gitter and we can route you to the right methods. I am locking this issue.
